### ISSUE TITLE
Text input with enter_returns_true doesn't commit full string

### ISSUE THREAD

**Issue by cosmicchipsocket** [imgui-rs/imgui-rs#564]
**Describe the bug**
If using a text input constructed with `enter_returns_true(true)`, pressing Enter to commit the string results in the string getting truncated to old length + 1 characters.

**To Reproduce**
```
let mut my_string = String::new();
ui.input_text("My input", &mut my_string).enter_returns_true(true).build();
```

**Expected behavior**
Full string is committed.

**Screenshots**
![TypingIssue](https://user-images.githubusercontent.com/34800072/141345303-1c7f777c-dbfc-4cf4-8939-2332f8bba297.gif)
![TypingCrash](https://user-images.githubusercontent.com/34800072/141345310-106e84e0-baed-4dc3-b36f-ba534fe92847.gif)

**Please describe your environment**
 - imgui-rs version 0.8.0
 - Windows 10

**Other information**
Sometimes typing enough into a box will cause a crash due to heap corruption!

**Comment by sanbox-irl** [comment]
Can you give this a try on Main? I believe we fixed this already

**Comment by cosmicchipsocket** [comment]
I will test it once I get finished porting my code to be up to speed with the latest API breakage. I thought it was only yesterday that the API was moving away from `ui.widget_name()` to `imgui::WidgetName::new()` and now in the next release it's reversing course. I recognize that the API needs to be cleaned up (in particular I appreciate that we can use Rust native strings now) but this is far too many changes too quickly, many of which seem to be wholly arbitrary. It's getting exhausting having to constantly update every single instance of every widget invocation. Now `imgui::Window::new(label).build(ui)` doesn't work, so I have to replace it with `imgui::Window::new(ui, label).build()`. Which by the way is not the same argument ordering as `imgui::MenuItem::new()` so I get to look forward to that breaking in the future too.

**Comment by cosmicchipsocket** [comment]
Seems to be fixed.

**Comment by cosmicchipsocket** [comment]
Would it be at all possible to get the text buffer fixes cherry-picked into a hotfix on crates.io? I'd like to be able to put out a fixed editor and continue working on my engine without having to deal with the current breakage in main.

**Comment by sanbox-irl** [comment]
I think I'm okay with a cherry picked 0.8.1 on crates.io. you're not the only one who has ran into this. 

We also have the issue with the free type feature, and I'd like to fix them both at once. I'll talk to Thom and get this sorted over the weekend

**Comment by cosmicchipsocket** [comment]
Nice! Thanks for the update, and sorry about the duplicate issue!

**Comment by cosmicchipsocket** [comment]
Hi, any news on this? I need to have working enter_returns_true, since otherwise every letter typed could trigger a package load in my editor which may create a pop-up error message, which would be even more disruptive than this bug.

**Comment by cosmicchipsocket** [comment]
Hi, I was wondering if there's anything I can do to help get this moving along faster? If an official hotfix release isn't made I might need to make my own crate with the fix. As much as I really dislike the idea of polluting the crates.io package namespace with a third party hotfix stopgap for a specific version of a specific crate, it would at least allow me to continue with my own work. It would also force consumers of imgui to point to the fork crate in Cargo.toml and modify their code to use the forked crate's name if they need to use it with another library that depends on said fork...

Might also want to look into getting an entry filed with rustsec/advisory-db and version 0.8.0 yanked since this is technically a heap corruption bug.

**Comment by dbr** [comment]
I've made at attempt at finding the relevant bugfixes from `main` (there was a lot more commits since 0.8 than I thought!) and have made #581

Could you give it a go and check it resolves the problem:

```toml
[patch.crates-io]
imgui = { git = "https://github.com/imgui-rs/imgui-rs.git", rev = "1159344386d42202db984d40f0e42c8b1716f836" }
imgui-sys = { git = "https://github.com/imgui-rs/imgui-rs.git", rev = "1159344386d42202db984d40f0e42c8b1716f836" }
```

**Comment by cosmicchipsocket** [comment]
Did some preliminary testing and it seems to be working. I can paste and commit text and it accepts the whole thing without crashing.

**Comment by dbr** [comment]
After a lot of delays, busy lives, and a confused bunch of 0.8 patch releases, these fixes should now definitely be out in v0.9, and future patch releases should be a lot smoother :sweat:
